### PR TITLE
nala: update to `0.11.1`

### PR DESCRIPTION
**Pull request by D-Brox**
I know this is a duplicate of 11337. I've already tried to make a PR with these updates on their fork, but it's been more than a month without new updates/commits. I've already discussed about updating the version with @volitank, the nala dev.

Please refer to #11337 for the original discussion around the patch improvements introduced in 0.11.0, aimed at termux.

### CODE REVIEW

**Comment by TomJo2000** [comment]
Feel free to take this, I have not had any time to work on this.
I'm sorry, I shouldn't have committed to taking this on since I knew I'd have limited time with my new internship but I thought I'd be able to work on it.

**Comment by D-Brox** [comment]
Nala is building correctly, but I overlooked that installing from pyproject does not generate a `.egg` file. I'll update the script and squash the commits.

**Comment by D-Brox** [comment]
I also noticed the build script uses `termux_setup_python_crossenv`. I'd like to ask if this step is really needed, considering that nala doesn't compile anything arch dependent.

**Comment by TomJo2000** [comment]
Could you also try to run this against [`nala 0.11.1`](https://gitlab.com/volian/nala/-/tree/v0.11.1)?
It'd be helpful to see if this script would continue to work in the future, and potentially be able to do auto-updates.

**Comment by D-Brox** [comment]
Nala 0.11.1 was a hot-fix for a debian-only issue, and has not been added to the volian apt repo, so I can't really test it.

But now the package should be updatable by changing only the version and hash, in theory.

**Comment by TomJo2000** [comment]
Ah, I hadn't looked into it that much.

LGTM. Thanks for your time.

**Comment by volitank** [comment]
@TomJo2000 

0.11.1 was only a small change to how we generate the date for the manpages during build time so that it is reproducible in Debian. This hot-fix changes nothing else for 0.11.0, the only thing is that the manpages will have the date from when it is built.

Considering that the termux build doesn't use the Debian packaging and instead uses pip there will be no differences at all.

Come to think of it though, it may be worth considering adding in `translations`, `manpages` and `shell-completions` to the install script.

for man pages these can be generated with `./nala_build.py man --install` and the translations with `./nala_build.py babel --compile --install`.

I don't know if these will put them in the right place, but it does use the termux path variable I added to make things easier. If not you can build them only by ommiting the `--install` switch, and then place them in the filesystem accordingly. If y'all end up trying this and install doesn't work, let me know and I'll update the build script for the next release.

For shell completions the files are Zsh: `debian/_nala`, Bash: `debian/bash-completion`, Fish: `debian/nala.fish`

Goal is to make Termux updates as painless as possible for everyone involved.

Also @D-Brox since they are using the GitLab Release page instead of my repo, all you have to do is change the version to `0.11.1` and it'll grab the source for that release.

**Comment by 2096779623** [comment]
Version `0.11.1` of nala has been released!

**Comment by D-Brox** [comment]
Sorry for

**Comment by D-Brox** [comment]
Also, could you answer https://github.com/termux/termux-packages/pull/11843#issuecomment-1239595889? Since it could simplify the build script even further if the function is not needed

**Comment by TomJo2000** [comment]
I just tested out the result from the latest CI run, it seems to work fine, but I had to manually create the `$PREFIX/var/lock/` directory, otherwise it would crash with this message about being unable to create a lock file.
```
Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/bin/nala", line 8, in <module>
    sys.exit(main())
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__main__.py", line 50, in main
    raise error from error
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__main__.py", line 41, in main
    nala()
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/typer/main.py", line 214, in __call__
    return get_command(self)(*args, **kwargs)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/typer/main.py", line 532, in wrapper
    return callback(**use_params)  # type: ignore
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/nala.py", line 236, in _update
    sudo_check()
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/utils.py", line 368, in sudo_check
    NALA_LOCK_FILE.touch(exist_ok=True)
  File "/data/data/com.termux/files/usr/lib/python3.10/pathlib.py", line 1168, in touch
    self._accessor.touch(self, mode, exist_ok)
  File "/data/data/com.termux/files/usr/lib/python3.10/pathlib.py", line 331, in touch
    fd = os.open(path, flags, mode)
FileNotFoundError: [Errno 2] No such file or directory: '/data/data/com.termux/files/usr/var/lock/nala.lock'
```
This might be a local issue on my end, but will need to be investigated before this can be merged

**Comment by D-Brox** [comment]
> This might be a local issue on my end, but will need to be investigated before this can be merged 

I guess the dir can be created in the postinst debscript, along with the other ones, of it doesn't exist. Tho that dir probably exists in most cases, since it's commonly used for lock files.

**Comment by TomJo2000** [comment]
> [I guess the dir can be created in the postinst debscript, along with the other ones, of it doesn't exist. Tho that dir probably exists in most cases, since it's commonly used for lock files.](https://github.com/termux/termux-packages/pull/11843#issuecomment-1244808481)

Yeah, as I said, probably a local issue, it would still be best to create the directory if it is missing to prevent such an issue.

File: packages/nala/build.sh
Comment (TomJo2000) [inline packages/nala/build.sh:42]:
```suggestion
	mkdir -p $TERMUX_PREFIX/var/log/nala
	mkdir -p $TERMUX_PREFIX/var/lock
```
Adresses crash described [here](https://github.com/termux/termux-packages/pull/11843#issuecomment-1244690489) related to `$TERMUX_PREFIX/var/lock/` (potentially) missing, thus being unable to create a lock file.

**Comment by D-Brox** [comment]
I was testing my patches directly on termux, seems like I need to add a pip dependency (`python-debian`) to`pyproject.toml`, as it's not being installed by default.

**Comment by D-Brox** [comment]
It's been almost a month, can someone review this please. At this rate Nala 0.12.0 will be released before this PR is merged.

Oh, and I'm still waiting for an answer about the use of `termux_setup_python_crossenv`

File: packages/nala/build.sh
Comment (Grimler91) [inline packages/nala/build.sh:33]: Better to use `python${TERMUX_PKG_VERSION}` and `pip${TERMUX_PKG_VERSION}` here

**Comment by Grimler91** [comment]
> I also noticed the old build script uses `termux_setup_python_crossenv`. I'd like to ask if this step is really needed, considering that nala doesn't compile anything arch dependent.

Python-crossenv is mainly used to allow cross-compilation of native libraries, so if nala doesn't have and native libraries it might not be needed. If it compiles and works fine without using python-crossenv then it can be dropped

File: packages/nala/build.sh
Comment (Grimler91) [inline packages/nala/build.sh:66]: Why is this removed? If there are python dependencies then they need to be installed through pip

Comment (Grimler91) [reply packages/nala/build.sh:66]: Was python-debian a runtime dependency? If yes then it should be installed here as well

**Review by Grimler91 — COMMENTED** [review]
Thanks for the PR! See comments/questions

File: packages/nala/build.sh
Comment (D-Brox) [reply packages/nala/build.sh:66]: If you install package with `pyproject.toml` with pip, all the dependencies will be installed alongside it.
I tested both in the termux build docker and installing using the patches on termux.
(Installing the runtime dependencies explicitly is still needed in packages that use legacy build formats like `setup.py` and `setup.cfg`)
image
image

Comment (D-Brox) [reply packages/nala/build.sh:66]: `python-debian` is a runtime dependency that @volitank forgot to add to the `pyproject.toml` file (since it's installed by default on debian-based distros), but will probably be on it in the next release.

File: packages/nala/build.sh
Comment (D-Brox) [reply packages/nala/build.sh]: In this case it would just be `python${_PYTHON_VERSION} -m pip (...)`, without pip version, since we are calling the module, not the executable

**Review by Grimler91 — CHANGES_REQUESTED** [review]
With the current build recipe the files are installed to `$TERMUX_PREFIX/local` instead of `$TERMUX_PREFIX` for some reason.

The package should only contain nala's files. As it is right now it contains all the python dependencies as well (h11, httpcore, python-debian, ...) which will lead to issues later. Dependencies should either be separate packages, or be installed through pip

**Comment by D-Brox** [comment]
> With the current build recipe the files are installed to `$TERMUX_PREFIX/local` instead of `$TERMUX_PREFIX` for some reason.

I have no idea why pip is doing that. I could use `pip install . --target=$PYTHONPATH` and move the executable to `$TERMUX_PREFIX/bin` if you prefer.

> The package should only contain nala's files. As it is right now it contains all the python dependencies as well (h11, httpcore, python-debian, ...) which will lead to issues later. Dependencies should either be separate packages, or be installed through pip

Ok, I will use `pip install . --no-deps` and install them with pip on the postinst script then.

**Comment by D-Brox** [comment]
Ok, I found the issue with `$TERMUX_PREFIX/local`:
- [Debian patches that change pip behaviour.](https://salsa.debian.org/cpython-team/python3/-/blob/39824a712442ff586962b565ac5ca50ccb25fea5/debian/patches/sysconfig-debian-schemes.diff)
- [Python discussion board about this matter.](https://discuss.python.org/t/linux-distro-patches-to-sysconfig-are-changing-pip-install-prefix-outside-virtual-environments/18240)

**Comment by Grimler91** [comment]
Using `termux_setup_python_crossenv` sounds like a better idea to me, a package shouldn't really modify the host files. If a user would abort the build then they would have a broken sysconfig.py file on their computer

**Comment by D-Brox** [comment]
> Using `termux_setup_python_crossenv` sounds like a better idea to me, a package shouldn't really modify the host files. If a user would abort the build then they would have a broken sysconfig.py file on their computer

Yes, I noticed that. I ended up using a virtualenv instead of the crossed, since it proved to be way faster  while achieving that same desired result.

**Comment by D-Brox** [comment]
@Grimler91 I think the update is now finally ready. 

What's been done:

- The patches were simplified following the modifications made by @volitank to increase termux compatibility.
- Since nala now uses a `pyproject.toml` instead of a `setup.py` file, the installation method was simplified and doesn't use the deprecated python `egg` as before.
- `pip` installs the deps from `pyproject.toml` packages by default, but those shouldn't be packaged inside the deb. The build script installs python deps with pip on the postinst script, as it was doing before.
- To circunvent the [distro dependent `sysconfig.py` issue](https://discuss.python.org/t/linux-distro-patches-to-sysconfig-are-changing-pip-install-prefix-outside-virtual-environments/18240), which makes `pip` install in `/local/lib/python3.X/dist-packages`, I used a dummy `virtualenv` to "trick" python into always installing in `/lib/python3.X/site-packages`.
The venv from `termux_setup_python_crossenv` could also have been used, but it proved to be slower.
- I also removed all unused variables, functions and patch files.

The next updates will probably not take this long, since they'll hopefully only change the version and hash (at least until the rust rewrite is released).

File: scripts/properties.sh
Comment (Grimler91) [inline scripts/properties.sh:42]: Please remove this from the commit

File: packages/nala/build.sh
Comment (Grimler91) [inline packages/nala/build.sh:30]: Seems `typing` (and potentially other dependencies) are missing here

File: packages/nala/build.sh
Comment (Grimler91) [inline packages/nala/build.sh:15]: Please re-add this so the `__pycache__` folder is removed from the package

**Review by Grimler91 — CHANGES_REQUESTED** [review]
Thanks for the changes, please update the commit message so that it is prefixed with `nala:`, so that it is easy to see which packages it is that is changed

File: scripts/properties.sh
Comment (D-Brox) [reply scripts/properties.sh:42]: Oh, sorry, I didn't notice that I added that to the commit when I was testing

File: packages/nala/build.sh
Comment (D-Brox) [reply packages/nala/build.sh:30]: There are only 2 "missing" dependencies:
- `typer` is automatically installed because its a dependency of `rich`.
- `typing-extentions`, which is a backport of `typing` features, only needed for python <3.10 when using nala, and wasn't added here because that's the version on the termux repos.

I'll add them anyway, since we may not know if the user is using the latest version of python

File: packages/nala/build.sh
Comment (D-Brox) [reply packages/nala/build.sh:15]: I'm pretty sure here is no need, since that folder is not being generated. Python packages build with `pyproject.toml` ensure build isolation by defauild, following PEP 518.

Comment (Grimler91) [reply packages/nala/build.sh:15]: Right, the pycache folder is in the nala subdirectory, so path to remove should be `lib/python${_PYTHON_VERSION}/site-packages/nala/__pycache__`:

```
$ dpkg -c output/nala_0.11.1_all.deb | grep pycache
drwx------ builder/builder     0 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/
-rw------- builder/builder 10037 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/cache.cpython-310.pyc
-rw------- builder/builder 16291 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/utils.cpython-310.pyc
-rw------- builder/builder 25206 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/install.cpython-310.pyc
-rw------- builder/builder 16919 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/error.cpython-310.pyc
-rw------- builder/builder  4465 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/constants.cpython-310.pyc
-rw------- builder/builder  1724 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/__init__.cpython-310.pyc
-rw------- builder/builder  9085 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/summary.cpython-310.pyc
-rw------- builder/builder  4388 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/search.cpython-310.pyc
-rw------- builder/builder  6939 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/debfile.cpython-310.pyc
-rw------- builder/builder 14295 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/nala.cpython-310.pyc
-rw------- builder/builder 21085 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/fetch.cpython-310.pyc
-rw------- builder/builder 13693 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/options.cpython-310.pyc
-rw------- builder/builder 18419 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/downloader.cpython-310.pyc
-rw------- builder/builder  6214 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/rich.cpython-310.pyc
-rw------- builder/builder 14657 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/history.cpython-310.pyc
-rw------- builder/builder  1053 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/__main__.cpython-310.pyc
-rw------- builder/builder 29691 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/dpkg.cpython-310.pyc
-rw------- builder/builder  9344 2022-10-23 09:13 ./data/data/com.termux/files/usr/lib/python3.10/site-packages/nala/__pycache__/show.cpython-310.pyc
```

Comment (D-Brox) [reply packages/nala/build.sh:15]: Ok, I fixed the removal path

**Comment by D-Brox** [comment]
@buttaface @Grimler91 @landfillbaby @2096779623
I'm taking the liberty to ping all termux reviewers as, for some reason, only Grimler91 is listed as a reviewer on this PR:
image

**Comment by Grimler91** [comment]
@xtkoba argued in https://github.com/termux/termux-packages/issues/12576#issuecomment-1291959034 that we should use termux_setup_python_crossenv for all python packages, so could you please add usage of that function again

**Comment by D-Brox** [comment]
@Grimler91 Anything else?

**Review by Grimler91 — APPROVED** [review]

**Comment by Grimler91** [comment]
> @Grimler91 Anything else?

Otherwise it looks good to me, if CI build succeeds and files are installed to correct place. Will merge in ~12 h unless someone else has comments, thanks

**Comment by Grimler91** [comment]
Thanks!

File: packages/nala/build.sh
Comment (D-Brox) [inline packages/nala/build.sh:16]: @Grimler91 shouldn't this 2 variables have been exported? Since they are now restricted to this function

Comment (Grimler91) [reply packages/nala/build.sh:16]: It works as is, basically our build.sh script is equivalent to:

```
$ cat build.sh
foo() {
    _PYTHON_VERSION=123
}
bar() {
    echo "something something $_PYTHON_VERSION"
}
```

and `build-package.sh` then runs

```
source build.sh
foo
bar
```

which echos `something something 123`, so `_PYTHON_VERSION` is set in bar() even though it is not exported in foo().

export would be needed if we create a new subshell after sourcing the build.sh, without export this echoes nothing:

```
source build.sh
foo
bash -c 'echo $_PYTHON_VERSION'
```

Comment (D-Brox) [reply packages/nala/build.sh:16]: Sometimes I forget how unintuitive bash can be sometimes

**Pull request merged**
